cdap-ui/app/features/datasets/controllers/tabs/query-model.js

```diff
@@ -0,0 +1,79 @@
+angular.module(PKG.name+'.feature.datasets')
+  .factory('QueryModel', function ($q, mySettings) {
+
+      function QueryModel(dataSrc, key, data) {
+        this.dataSrc = dataSrc;
+        this.key = key;
+        this.data = data || [];
+        if (this.data.length) {
+          this.sync();
+        }
+      }
+
+      QueryModel.prototype.add = function (entry) {
+        if(this.data.indexOf(entry) == -1) {
+          this.data.push(entry);
+          this.sync();
+        }
+      };
+
+      QueryModel.prototype.set = function (entries) {
+        if(angular.isArray(entries)) {
+          this.data = entries;
+          this.sync();
+        }
+      };
+
+      QueryModel.prototype.get = function () {
+        var self = this;
+        var deferred = $q.defer();
+        var requests = [];
+        requests.push(mySettings.get(this.key));
+        requests.push(this.dataSrc.request({
+          _cdapNsPath: '/data/explore/queries',
+          })
+        );
+        $q.all(requests).then(function (responses) {
+          self.data = self.mergeQueries(responses[0], responses[1]);
+          deferred.resolve(self.data);
+          self.sync();
+        });
+        return deferred.promise;
+      };
+
+      /**
+       * Merges queries stored in user settings and queries pulled from server
+       * using query_handle.
+       */
+      QueryModel.prototype.mergeQueries = function (userQueries, apiQueries) {
```

Comment:
We get the query handles from two endpoints, 

- From backend: We get all the queries made by all the users 
- From user store: We get a list of query handles for the current user .

Some of the query handles that we had stored in the backend might have expired. So at time t=0th day backend might have sent 30 queries and the current user might have made 5 queries out of that.
So now we should show only those 5 queries in the UI.

And lets say the query expiration time is 1 day and after a day (time t=1day) we query for list of queries we might not get all the 30 because from the backend since some might have expired.  
When we say the queries expired it might be queries made by the current user + queries made by other users.

So now we need to check if the list of query handles that we get from user store, against the list of query handles we get from backend, are valid and are not expired and show only those that belong to the current user. 

Hope I am making some sense here.